Replace sync and cron-based EA Q processing system with async

lucatume opened the pull request:
Ticket: #105650

Based on: https://github.com/moderntribe/tribe-common/pull/690

This PR replaces the default EA Client queue processing system with an async one.
The previous one is left in place and simply used when forced.
On an high level, due to its async nature, this is how it works:
1. parse all the items received from the service: if any linked post used in an item is used in another item too (e.g. 2 events with the same venue); then the second item become *dependent* on the linked post defined in the first item
2. any event (item) that has no linked posts (venue, org) is immediately inserted; any event that has linked posts **and**, in phase 1, was marked as *dependent* is moved to the back of the queue if the linked posts it depends on do not exist yet
2bis. if an event comes with linked post information and is not *dependent* then it's immediately inserted and an import-wise transitional meta is set on the linked posts; *dependent* events will check that to make sure they can be created

All the above to avoid critical runs in linked posts, we know, from the service, that all events are unique

tr1b0t commented:
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $batch_size is never used | [Async_Queue.php:163](https://github.com/moderntribe/the-events-calendar/blob/e860350326d70d439ef69bab8fa2c745cb4863e7/src/Tribe/Aggregator/Record/Async_Queue.php#L163)
:x: | Expected 1 space after closing parenthesis; found 0 | [Items.php:38](https://github.com/moderntribe/the-events-calendar/blob/e860350326d70d439ef69bab8fa2c745cb4863e7/src/Tribe/Aggregator/Record/Items.php#L38)

> via jenkins/codesniffer-phpcs

tr1b0t commented:
Closing this Pull Request because it is pull requested against the wrong branch. The Central ticket is for *1.0.20*, but this was pull requested against *release/M18.07*.

tr1b0t commented:
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $batch_size is never used | [Async_Queue.php:163](https://github.com/moderntribe/the-events-calendar/blob/a6af2396d621f6e4fd1d1dec8adc21e94911deb0/src/Tribe/Aggregator/Record/Async_Queue.php#L163)

> via jenkins/codesniffer-phpcs

sc0ttkclark commented on src/Tribe/Venue.php:
```diff
@@ -451,14 +451,6 @@ public function create( $data, $post_status = 'publish', $avoid_duplicates = fal
 
 				$this->save_meta( $venue_id, $data );
 
-				/**
```
Intentional?

lucatume replied:
Yes, intentional. The action would be fired twice with the same parameters.
My guess is that was the result of a merge.

tr1b0t commented:
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $batch_size is never used | [Async_Queue.php:163](https://github.com/moderntribe/the-events-calendar/blob/f4ae80a3101e95ef417076ff96113f11a1d8d630/src/Tribe/Aggregator/Record/Async_Queue.php#L163)

> via jenkins/codesniffer-phpcs

tr1b0t commented:
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $batch_size is never used | [Async_Queue.php:163](https://github.com/moderntribe/the-events-calendar/blob/bcbe8e23a97848bef26002811f1c257641aa1d6c/src/Tribe/Aggregator/Record/Async_Queue.php#L163)

> via jenkins/codesniffer-phpcs

tr1b0t commented:
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $batch_size is never used | [Async_Queue.php:163](https://github.com/moderntribe/the-events-calendar/blob/bcbe8e23a97848bef26002811f1c257641aa1d6c/src/Tribe/Aggregator/Record/Async_Queue.php#L163)

> via jenkins/codesniffer-phpcs

tr1b0t commented:
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $batch_size is never used | [Async_Queue.php:171](https://github.com/moderntribe/the-events-calendar/blob/f5a9cb1f225e63f427652d9f138d1dc72c60c92e/src/Tribe/Aggregator/Record/Async_Queue.php#L171)
:warning: | The method parameter $ignored is never used | [CSV.php:119](https://github.com/moderntribe/the-events-calendar/blob/f5a9cb1f225e63f427652d9f138d1dc72c60c92e/src/Tribe/Aggregator/Record/CSV.php#L119)

> via jenkins/codesniffer-phpcs

lucatume commented:
I've rebased the PR on the `release/M18.07.1` branch and fixed a number of issues related to the use of the `wp event-aggregator` wp-cli command and the new queueing system.

tr1b0t commented:
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $batch_size is never used | [Async_Queue.php:171](https://github.com/moderntribe/the-events-calendar/blob/584303f348b18beebae06bc0350c33874917ac93/src/Tribe/Aggregator/Record/Async_Queue.php#L171)
:warning: | The method parameter $ignored is never used | [CSV.php:119](https://github.com/moderntribe/the-events-calendar/blob/584303f348b18beebae06bc0350c33874917ac93/src/Tribe/Aggregator/Record/CSV.php#L119)

> via jenkins/codesniffer-phpcs

barryhughes commented on src/Tribe/Aggregator/Cron.php:
```diff
@@ -567,4 +569,22 @@ public function purge_expired_records() {
+	 *
+	 * @since TBD
+	 *
+	 * @param $Tribe__Events__Aggregator__Record__Abstract $record
```
Minor typo - `Tribe__Events__Aggregator__Record__Abstract` shouldn't be prefixed with a `$` sigil

barryhughes commented on src/Tribe/Aggregator/Processes/Import_Events.php:
```diff
@@ -0,0 +1,365 @@
+	 * @return Tribe__Events__Aggregator__Processes__Import_Events
+	 */
+	public function save() {
+		add_filter( "tribe_process_queue_{$this->identifier}_save_data", array( $this, 'save_data' ) );
```
With this code sitting inside of the EA client, is that the [correct prefix](http://moderntribe.github.io/products-engineering/guidelines/plugins-namespacing/) (should it be `tribe_aggregator_`)?

barryhughes replied:
...ditto other hooks in this class

sc0ttkclark replied:
This is an add_filter, the hook itself is running in tribe-common

barryhughes commented on src/Tribe/Aggregator/Processes/Import_Events.php:
```diff
@@ -0,0 +1,365 @@
+
+		if ( empty( $dependencies ) ) {
+			$this->has_dependencies = false;
+			$activity               = $this->insert_event( $record_id, (object) $data );
```
Doesn't `insert_event()` expect an array rather than an object (or perhaps we need to update the docblock for that method)? Same question applies to line 188.

lucatume replied:
The doc-block needed to be updated, it expects an object, in fact.

barryhughes commented on src/Tribe/Aggregator/Processes/Import_Events.php:
```diff
@@ -0,0 +1,365 @@
+	 * @param int $record_id
+	 * @param array $data
+	 *
+	 * @return Tribe__Events__Aggregator__Record__Activity
```
Can also return bool (line 232)

barryhughes commented on src/Tribe/Aggregator/Processes/Import_Events.php:
```diff
@@ -0,0 +1,365 @@
+	 *
+	 * @param int $record_id
+	 *
+	 * @return null|Tribe__Error|Tribe__Events__Aggregator__Record__Abstract
```
I don't think it can return Tribe__Error (I feel like this mistake has kind of spread virally across the plugin code) ... Tribe__Error is used to generate WP_Error objects but generally instances of Tribe__Error aren't passed around.

bordoni commented on src/Tribe/Aggregator/Processes/Service_Provider.php:
```diff
@@ -0,0 +1,36 @@
+		tribe_register( 'events-aggregator.record-items', 'Tribe__Events__Aggregator__Record__Items' );
+		tribe_register( 'events-aggregator.processes.import-events', 'Tribe__Events__Aggregator__Processes__Import_Events' );
+
+		add_filter( 'tribe_process_queues', array( $this, 'filter_tribe_process_queues' ) );
```
Should we move the hook's into another method?

lucatume replied:
not right now, I'm always for clean code but creating a method to call another for no added readability effect seems like speculative coding.
If there were two, though, I would have done it.

lucatume commented on src/Tribe/Aggregator/Record/Async_Queue.php:
```diff
@@ -0,0 +1,261 @@
+	 *
+	 * @return Tribe__Events__Aggregator__Record__Async_Queue
+	 */
+	public function process( $batch_size = null ) {
```
There is no return from `dispatch` as it's async. This is the typical "tell, don't ask" method: we rely on the method to setup its own chain of effects.
I've modified the method to return, as its doc-block would imply, the queue object itself.

lucatume commented:
Put in place the requested fixes, ready for another look.

tr1b0t commented:
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $batch_size is never used | [Async_Queue.php:171](https://github.com/moderntribe/the-events-calendar/blob/379fc44b98fc1c7c01f8fff0cb2d1fa4e308703d/src/Tribe/Aggregator/Record/Async_Queue.php#L171)
:warning: | The method parameter $ignored is never used | [CSV.php:119](https://github.com/moderntribe/the-events-calendar/blob/379fc44b98fc1c7c01f8fff0cb2d1fa4e308703d/src/Tribe/Aggregator/Record/CSV.php#L119)

> via jenkins/codesniffer-phpcs
